Lib/fractions.py

```diff
@@ -564,10 +564,25 @@ def __hash__(self):
         try:
             dinv = pow(self._denominator, -1, _PyHASH_MODULUS)
         except ValueError:
-            # ValueError means there is no modular inverse
+            # ValueError means there is no modular inverse.
             hash_ = _PyHASH_INF
         else:
-            hash_ = hash(abs(self._numerator)) * dinv % _PyHASH_MODULUS
+            # The general algorithm now specifies that the absolute value of
+            # the hash is
+            #    (|N| * dinv) % P
+            # where N is self._numerator and P is _PyHASH_MODULUS.  That's
+            # optimized here in two ways:  first, for a non-negative int i,
+            # hash(i) == i % P, but the int hash implementation doesn't need
+            # to divide, and is faster than doing % P explicitly.  So we do
+            #    hash(|N| * dinv)
+            # instead.  Second, N is unbounded, so its product with dinv may
+            # be arbitrarily expensive to compute.  The final answer is the
+            # same if we use the bounded |N| % P instead, which can again
+            # be done with an int hash() call.  If 0 <= i < P, hash(i) == i,
+            # so this nested hash() call wastes a bit of time making a
+            # redundant copy when |N| < P, but can save an arbitrarily large
+            # amount of computation for large |N|.
+            hash_ = hash(hash(abs(self._numerator)) * dinv)
```

Comment:
Interesting that
```python
hash_ = hash(hash(abs(self._numerator)) * dinv)
result = hash_ if self._numerator >= 0 else -hash_
```
is equal to
```python
result = hash(hash(self._numerator) * dinv)
```
except the cases when `self._numerator` is -1 (or `-1 - _PyHASH_MODULUS*k` where `k` is a non-negative integer).

Reply:
It's a just a consequence of that, for integer `i`,
```python
    hash_ = hash(abs(i))
    result = hash_ if i >= 0 else -hash_
```
is the same as
```python
    result = hash(i)
```
unless `i` is congruent to -1 mod P.  Which has nothing to do with "the math", but instead with that Python doesn't allow -1 as a hash result.